Add pagination for submissions

Koc opened the pull request:
This PR adds pagination for submissions list page. Few notes about implementation:
- It's not affect "Summary" tab
- Pagination works on the server
- There is an extra "search" field that used for filtering submissions

<details><summary>:mag: Preview</summary>
<p>

![image](https://github.com/user-attachments/assets/f5318431-ebe1-4830-a66e-a2148bf4b2b7)
![image](https://github.com/user-attachments/assets/892cadd9-ae7c-4307-beb2-b58dd36a1fc2)


</p>
</details>

Koc commented on src/views/Results.vue:
```diff
@@ -201,12 +201,79 @@
 
 			<!-- Responses view for individual responses -->
 			<section v-else>
+				<div class="pagination-block" :class="{ 'large-width': isMobile }">
```
Stolen from tables :ninja: . In theory we can extract this as a separate component to UI library

Chartman123 reviewed (commented):
Hi @Koc 

nice start, here are a few comments:

- I think we should split up the filtering from the rest of the pagination block and move it up to the right of the summary/responses switch
![grafik](https://github.com/user-attachments/assets/dbdfa432-bd22-4b65-ba4b-fad5a0851bd8)
- The page dropdown is placed too high, there should be more spacing on top of it
- `floating` style looks odd to me, should be just embedded into the page
- page selection should be below the submissions or above and below but not only above them.
- A max of 50 items seems too big for me as a default. You'd still have to scroll a lot before you reach the end of the page
- At best we should make the item limit configurable (user setting)

Regarding the pagination in general: I think we should only fetch the currently shown submissions from the API. But in this case the filtering will no longer be able to work only in the front-end. So more adjustments needed then...

Koc commented on src/views/Results.vue:
```diff
@@ -201,12 +201,79 @@
+						<NcButton
+							type="tertiary"
+							:disabled="totalPages === 1 || pageNumber <= 1"
+							:aria-label="t('tables', 'Go to first page')"
```
copy-paste detected :see_no_evil:

Koc commented on src/views/Results.vue:
```diff
@@ -348,6 +427,10 @@ export default {
 
+			submissionSearch: '',
+			pageNumber: 1,
+			rowsPerPage: 50,
```
set to 20 by default

Koc commented:
Hey @Chartman123 , thank you for quick feedback, very appreciated!

I've did some changes and updated screenshots in the PR description. So, I've moved pagination to backend and added extra parameter for full text search. For some reason it not works for me :confused: . Please help, feel free to push.

Koc commented on lib/Db/SubmissionMapper.php:
```diff
@@ -106,40 +113,61 @@ public function hasFormSubmissionsByUser(Form $form, string $userId): bool {
 		$result->closeCursor();
 
-		return (int)($row['num_submissions'] ?? 0);
+		return $result->rowCount();
```
this not works in SQLite :confused:

Chartman123 replied:
This can be fixed with my commit :)

Chartman123 commented on lib/Controller/ApiController.php:
```diff
@@ -1161,9 +1164,11 @@ public function getSubmissions(int $formId, ?string $fileFormat = null): DataRes
 		// Load submissions and currently active questions
 		if (in_array(Constants::PERMISSION_RESULTS, $this->formsService->getPermissions($form))) {
-			$submissions = $this->submissionService->getSubmissions($formId);
+			$submissions = $this->submissionService->getSubmissions($formId, null, $query, $limit, $offset);
+			$totalSubmissionsCount = $this->submissionMapper->countSubmissions($formId, null, $query);
```
Can be removed
```suggestion
			$submissions = $this->submissionService->getSubmissions($formId, null, $query, $limit, $offset);
```

Chartman123 commented on docs/API_v3.md:
```diff
@@ -741,7 +747,8 @@ Get all Submissions to a Form
       "extraSettings": {}
     }
-  ]
+  ],
+  "totalSubmissionsCount": 40
```
Can be reverted, is now part `GET /forms/{formId}
```suggestion
  ]
```

Chartman123 commented on lib/Controller/ApiController.php:
```diff
@@ -1161,9 +1164,11 @@ public function getSubmissions(int $formId, ?string $fileFormat = null): DataRes
+			$totalSubmissionsCount = $this->submissionMapper->countSubmissions($formId, null, $query);
 		} else {
-			$submissions = $this->submissionService->getSubmissions($formId, $this->currentUser->getUID());
+			$submissions = $this->submissionService->getSubmissions($formId, $this->currentUser->getUID(), $query, $limit, $offset);
+			$totalSubmissionsCount = $this->submissionMapper->countSubmissions($formId, null, $query);
```
Same here:
```suggestion
			$submissions = $this->submissionService->getSubmissions($formId, $this->currentUser->getUID(), $query, $limit, $offset);
```

Chartman123 commented on lib/Controller/ApiController.php:
```diff
@@ -1196,6 +1201,7 @@ public function getSubmissions(int $formId, ?string $fileFormat = null): DataRes
 		$response = [
 			'submissions' => $submissions,
 			'questions' => $questions,
+			'totalSubmissionsCount' => $totalSubmissionsCount,
```
and here

Chartman123 commented on lib/ResponseDefinitions.php:
```diff
@@ -77,7 +77,8 @@
  * @psalm-type FormsSubmissions = array{
  *   submissions: list<FormsSubmission>,
- *   questions: list<FormsQuestion>
+ *   questions: list<FormsQuestion>,
+ *   totalSubmissionsCount: int
```
Same here
```suggestion
 *   questions: list<FormsQuestion>
```

Chartman123 commented on openapi.json:
```diff
@@ -592,7 +592,8 @@
                 "required": [
                     "submissions",
-                    "questions"
+                    "questions",
+                    "totalSubmissionsCount"
```
And here:
```suggestion
                    "questions"
```

Chartman123 commented on openapi.json:
```diff
@@ -606,6 +607,10 @@
                         "items": {
                             "$ref": "#/components/schemas/Question"
                         }
+                    },
+                    "totalSubmissionsCount": {
+                        "type": "integer",
+                        "format": "int64"
                     }
```
Can be removed

Chartman123 commented on src/views/Results.vue:
```diff
@@ -478,22 +544,25 @@ export default {
 				// Append questions & submissions
-				this.submissions = loadedSubmissions
-				this.questions = loadedQuestions
+				this.submissions = this.formatDateAnswers(
+					data.submissions,
+					data.questions,
+				)
+				this.questions = data.questions
+				this.totalSubmissionsCount = data.totalSubmissionsCount
```
```suggestion
				this.submissions = this.formatDateAnswers(
					data.submissions,
					data.questions,
				)
				this.questions = data.questions
```

Chartman123 commented on tests/Integration/Api/ApiV3Test.php:
```diff
@@ -1116,7 +1116,8 @@ public function dataGetSubmissions() {
 						]
 					],
-					'questions' => $this->dataGetFullForm()['getFullForm']['expected']['questions']
+					'questions' => $this->dataGetFullForm()['getFullForm']['expected']['questions'],
+					'totalSubmissionsCount' => 3,
```
```suggestion
					'questions' => $this->dataGetFullForm()['getFullForm']['expected']['questions']
```

Chartman123 commented on tests/Integration/Api/ApiV3Test.php:
```diff
@@ -1236,6 +1237,7 @@ public function testExportToCloud() {
 	public function dataDeleteSubmissions() {
 		$submissionsExpected = $this->dataGetSubmissions()['getSubmissions']['expected'];
 		$submissionsExpected['submissions'] = [];
+		$submissionsExpected['totalSubmissionsCount'] = 0;
```
Can be removed

Chartman123 commented on tests/Integration/Api/ApiV3Test.php:
```diff
@@ -1368,6 +1370,7 @@ public function dataDeleteSingleSubmission() {
 	 * @param array $submissionsExpected
 	 */
 	public function testDeleteSingleSubmission(array $submissionsExpected) {
+		$submissionsExpected['totalSubmissionsCount'] = $submissionsExpected['totalSubmissionsCount'] - 1;
```
Can be removed

Chartman123 commented on tests/Unit/Controller/ApiControllerTest.php:
```diff
@@ -245,6 +245,7 @@ public function dataGetSubmissions() {
 							'extraSettings' => new \stdClass(),
 						],
 					],
+					'totalSubmissionsCount' => 1,
```
Can be removed

Chartman123 commented on tests/Unit/Controller/ApiControllerTest.php:
```diff
@@ -265,6 +266,7 @@ public function dataGetSubmissions() {
 							'extraSettings' => new \stdClass(),
 						],
 					],
+					'totalSubmissionsCount' => 1,
```
Can be removed

Chartman123 commented on tests/Unit/Controller/ApiControllerTest.php:
```diff
@@ -298,6 +300,11 @@ public function testGetSubmissions(array $submissions, array $questions, array $
 			->with(1)
 			->willReturn($questions);
 
+		$this->submissionMapper->expects($this->once())
+			->method('countSubmissions')
+			->with(1)
+			->willReturn(1);
+
```
Can be removed

Chartman123 commented on src/views/Results.vue:
```diff
@@ -344,13 +396,18 @@ export default {
 
 			questions: [],
 			submissions: [],
+			totalSubmissionsCount: 0,
```
Can be removed

Chartman123 commented on src/views/Results.vue:
```diff
@@ -202,15 +235,27 @@
+
+				<PaginationToolbar
+					class="bottom-pagination"
+					:total-items-count="totalSubmissionsCount"
```
```suggestion
					:total-items-count="form.submissionCount"
```

Chartman123 reviewed (commented):
submissionCount is now part of the Form, you can just rely on it. I needed it for the allowEditSubmission PR :)

Koc commented:
@Chartman123 mate, the problem is that we have 2 different counts.

![image](https://github.com/user-attachments/assets/5fbe1b52-df05-420d-9f27-096639201282)

Chartman123 commented:
@Koc ah yes you're right... 🙈

Chartman123 commented:
But your count is wrong at the moment for shared forms:

![grafik](https://github.com/user-attachments/assets/462bf748-d791-4346-a945-6508fc959007)

You're counting all submissions and not only those by the current user if the user doesn't have the see all results permission

Chartman123 commented:
Count looks good now :) But I'm not happy with the name `totalSubmissionsCount` as it's not the total of all submissions available to the user. What about renaming it to `filteredSubmissionsCount`?

And the search box is not properly aligned to the other items in the block. You could also use the `var(--default-grid-baseline)` for the margin instead of the `4px`.

Koc commented:
@Chartman123 done :heavy_check_mark:

Chartman123 commented on src/views/Results.vue:
```diff
@@ -869,4 +938,13 @@ export default {
 }
+
+.search-wrapper {
+	margin-top: calc(-1 * var(--default-grid-baseline));
+	margin-left: var(--default-grid-baseline);
```
Let's change this to margin-inline-start/end and margin-block-start so that it also works in RTL mode. `auto` makes it aligned on the right and the end margin is needed because the header and the question sections are not of the same width.

```suggestion
	margin-block-start: calc(-1 * var(--default-grid-baseline));
	margin-inline-start: auto;
	margin-inline-end: var(--default-clickable-area);
```

Chartman123 reviewed (approved):
Two more little comments, but rest looks fine for me :)

I posted this PR to Design team chat and asked if someone could have a look at it design-wise...

jancborchardt reviewed (commented):
Looks good! Just 2 comments:

Usually the pagination element is on the bottom of the oage rather than the top. It can be made sticky on the bottom if wanted.

> At best we should make the item limit configurable (user setting)

@Chartman123 just for info – something unimportant as this should not be a setting but instead have a proper value, as per design guidelines. :)

Chartman123 commented:
@jancborchardt Thanks for the review 👍🏻 

> Usually the pagination element is on the bottom of the oage rather than the top.

@Koc could you then please remove the page selection from above the results?

Koc commented:
@Chartman123 removed
